internal/api/token.go

```diff
@@ -577,6 +589,95 @@ func (a *API) IdTokenGrant(ctx context.Context, w http.ResponseWriter, r *http.R
 	return sendJSON(w, http.StatusOK, token)
 }
 
+func (a *API) OAuthPKCE(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+	db := a.db.WithContext(ctx)
+	var grantParams models.GrantParams
+
+	params := &PKCEGrantParams{}
+	body, err := getBodyBytes(r)
+	if err != nil {
+		return internalServerError("Could not read body").WithInternalError(err)
+	}
+
+	if err = json.Unmarshal(body, params); err != nil {
+		return badRequestError("invalid body: unable to parse JSON").WithInternalError(err)
+	}
+
+	if params.AuthCode == "" || params.CodeVerifier == "" {
+		return badRequestError("invalid request: both auth code and code verifier should be non-empty")
+	}
+
+	flowState, err := models.FindFlowStateByAuthCode(db, params.AuthCode)
+	// Sanity check in case user ID was not set properly
+	if models.IsNotFoundError(err) || flowState.UserID == uuid.Nil {
+		return forbiddenError("invalid oauth state, please ensure oauth redirect has successfully completed")
+	} else if err != nil {
+		return err
+	}
+	flowStateExpiryDuration := a.config.External.FlowStateExpiryDuration
+	flowStateCreationTime := flowState.CreatedAt
+	flowStateExpiryTime := flowStateCreationTime.Add(time.Second * time.Duration(flowStateExpiryDuration))
+	if time.Now().After(flowStateExpiryTime) {
+		return forbiddenError("invalid oauth state, oauth state has expired")
+	}
+
+	user, err := models.FindUserByID(db, flowState.UserID)
+	if err != nil {
+		return err
+	}
+
+	if flowState.AuthCode != params.AuthCode {
+		return forbiddenError("invalid auth code")
+	}
```

Comment:
Is this a real case? When would the flow state not have the correct auth code?

Reply:
Yeah, you're right it's a no-op since `models.FindFlowStateByAuthCode(db, params.AuthCode)` would return a modelNotFound Error if the Auth Code does not match - let's remove